substrate-archive/src/database/batch.rs

```diff
@@ -19,16 +19,24 @@
 //! This is sort of temporary until SQLx develops their dynamic query builder: https://github.com/launchbadge/sqlx/issues/291
 //! and `Quaint` switches to SQLx as a backend: https://github.com/prisma/quaint/issues/138
 
+use async_std::{future::timeout, task};
+use futures::{future, stream, StreamExt, TryFutureExt, TryStreamExt};
 use sqlx::{
 	encode::Encode,
-	postgres::{PgArguments, PgConnection, Postgres},
+	postgres::{PgArguments, PgConnection, PgPool, Postgres},
 	prelude::*,
-	Arguments,
+	Arguments, Executor,
 };
 
-use crate::error::Result;
+use std::{
+	convert::TryInto,
+	sync::atomic::{AtomicUsize, Ordering},
+	time::Duration,
+};
+
+use crate::error::{ArchiveError, Result};
 
-const CHUNK_MAX: usize = 30_000;
+const CHUNK_MAX: usize = 5_000;
```

Comment:
Mostly picked from my own testing. It is dependent on the users hardware to some extent. It might be worth making it configurable from the TOML, however for the general case I think it is OK to give a hard-coded value, at least for now.

I chose it by running archive and checking how long it took to insert storage into Postgres. The insertion times were a big hangup in previous iterations, which was one of the causes of the tasks timing out. (There would be ~100-200 blocks to insert which each had up to a thousand storage changes each)

This was fixed by lowering `CHUNK_MAX` , but allowing storage to be inserted concurrently based on how many free idle connections to Postgres exist. This way instead of one query taking ~10-20s to insert 30K items, we can have 3-4 or more which each take less time, based on how many threads exist on the system. Insertion times for blocks have never really been an issue, so this is mostly an optimization for storage (storage contains all events, trie values, etc resulting from the execution of the extrinsics in the block).

Reply:
FWIW I for one would appreciate reading this whole comment directly in the code. I don't think comments must be super terse and I think it's fine to include the reasoning behind things, even if it's just to say "this is something of a guesstimate!".
Either way, ty for explaining!

Reply:
definitely noted. I'm oftentimes hesitant or forget to fully flesh out my thinking in comments, but i will definitely keep this in mind! Sometimes it's also difficult to tell what deserves to be explained further.

I did copy over a version of this for CHUNK_MAX though